File: python/779.py
"""Project Euler Problem 779: Prime Factor and Order.

Let f_K(n) = (α(n)-1)/p(n)^K, where p(n) is the smallest prime dividing
n and α(n) is its p-adic order. Find Σ_{K=1}^∞ lim_{N->∞} 1/N
Σ_{n=2}^N f_K(n).

Consider any p(n) and a(n). The fraction of integers with that p(n) and
a(n) is 1/p^a (the integers divisible by p^a) times (p-1)/p (the integers
not divisible by p^{a+1}) times Π_{q<p} (q-1)/q (the integers without
any smaller prime factors). So this contributes f_K(n) 1/p^a (p-1)/p
Π_{q<p} (q-1)/q to the sum. Summing this over all a and K gives:

S_p = Σ_a Σ_K (a-1)/p^K 1/p^a (p-1)/p Π_{q<p} (q-1)/q
    = Σ_a (a-1) 1/p^{a+1} Π_{q<p} (q-1)/q
    = 1 / (p(p-1)²) Π_{q<p} (1-1/q).

Summing this for enough p gives the answer.
"""

from __future__ import annotations

from math import isqrt
from typing import List


def sieve_primes(limit: int) -> List[int]:
    """Sieve of Eratosthenes."""
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False
    for i in range(2, isqrt(limit) + 1):
        if is_prime[i]:
            for j in range(i * i, limit + 1, i):
                is_prime[j] = False
    return [i for i in range(2, limit + 1) if is_prime[i]]


def solve() -> str:
    """Solve Problem 779."""
    L = 100_000
    primes = sieve_primes(L)
    ans = 0.0

    for i, p in enumerate(primes):
        res = 1.0 / p / ((p - 1) ** 2)
        for j in range(i):
            res *= 1 - 1.0 / primes[j]
        ans += res

    return f"{ans:.12f}"


def main() -> str:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/779.c
/*
 * Project Euler 779 - Prime Factor and Order
 *
 * Compute sum over primes p of 1/(p*(p-1)^2) * product_{q<p} (1-1/q).
 * Uses sieve of Eratosthenes for primes up to 100000.
 */
#include <stdio.h>
#include <math.h>

#define LIMIT 100000

static char is_prime[LIMIT + 1];
static int primes[10000];
static int nprimes;

static void sieve(void) {
    for (int i = 0; i <= LIMIT; i++) is_prime[i] = 1;
    is_prime[0] = is_prime[1] = 0;
    for (int i = 2; (long long)i * i <= LIMIT; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= LIMIT; j += i)
                is_prime[j] = 0;
        }
    }
    nprimes = 0;
    for (int i = 2; i <= LIMIT; i++)
        if (is_prime[i])
            primes[nprimes++] = i;
}

int main(void) {
    sieve();

    double ans = 0.0;
    double prod = 1.0; /* product of (1-1/q) for q < p */

    for (int i = 0; i < nprimes; i++) {
        double p = primes[i];
        double res = prod / (p * (p - 1.0) * (p - 1.0));
        ans += res;
        prod *= (1.0 - 1.0 / p);
    }

    printf("%.12f\n", ans);
    return 0;
}
